lib/Infer/ExhaustiveSynthesis.cpp

```diff
@@ -374,8 +377,12 @@ ExhaustiveSynthesis::synthesize(SMTLIBSolver *SMTSolver,
                                 const std::vector<InstMapping> &PCs,
                                 Inst *LHS, Inst *&RHS,
                                 InstContext &IC, unsigned Timeout) {
-  std::vector<Inst *> Inputs;
-  findCands(LHS, Inputs, /*WidthMustMatch=*/false, /*FilterVars=*/false, MaxLHSCands);
+  std::vector<Inst *> Vars;
+  findVars(LHS, Vars);
+
```

Comment:
we should talk about this part. why do you think it's a good idea? do you have data supporting that it does a better job?

Reply:
I did this because the current findCands() stops when it reaches MaxLHSCands(which is set to 15 by default). However, this might miss the input variable declared at the top of program. Due to this limitation, the synthesizer cannot gives correct ctpop result since the input %x is not harvested as a component of synthesis.

Exhaustive synthesis has a call findVars() after doing the big query, hoisting out this call does not have much impact to the compilation performance, compared to the potential benefit of getting better synthesis results. 

This approach is stupid I admit, it basically runs the traversal on the LHS twice. I can write a better version of the the findCands which runs once.

Reply:
This code basically guarantees all the program inputs are listed as a components in the synthesis.

Reply:
Okay this makes sense. But instead of stopping at a hard bounded limit when doing traversal, can we traverse all the tree and hard bounds the result by the descend order of the "benefit" of each candidates? There is already a "benefit" tagging algorithm there implemented in findVars.

Reply:
Currently the deeper the traverse goes, the higher benefit the candidate gets. We may further customize this benefit function to fit our needs. Say some algorithm can tell some kind of candidates are more likely to be used as a component in synthesis, then we can increase the "benefit" of this kind of candidates.

Reply:
Sure I will split this pr and collect data